nav2_behavior_tree/plugins/control/pause.cpp

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <thread>
+
+// ROS includes
+#include "nav2_behavior_tree/plugins/control/pause.hpp"
+#include "rclcpp/callback_group.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+// Other includes
+#include "behaviortree_cpp/bt_factory.h"
+
+namespace nav2_behavior_tree
+{
+
+using namespace std::placeholders;
+
+Pause::Pause(
+  const std::string & xml_tag_name,
+  const BT::NodeConfiguration & conf)
+: BT::ControlNode(xml_tag_name, conf)
+{
+  node_ = this->config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+  state_ = UNPAUSED;
+
+  // Create a separate cb group with a separate executor to spin
+  cb_group_ = node_->create_callback_group(
+    rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+  executor_ =
+    std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+
+  executor_->add_callback_group(cb_group_, node_->get_node_base_interface());
+
+  std::string pause_service_name;
+  getInput("pause_service_name", pause_service_name);
+  pause_srv_ = node_->create_service<Trigger>(
+    pause_service_name,
+    std::bind(&Pause::pause_service_callback, this, _1, _2),
+    rclcpp::ServicesQoS(), cb_group_);
+
+  std::string resume_service_name;
+  getInput("resume_service_name", resume_service_name);
+  resume_srv_ = node_->create_service<Trigger>(
+    resume_service_name,
+    std::bind(&Pause::resume_service_callback, this, _1, _2),
+    rclcpp::ServicesQoS(), cb_group_);
+
+  spinner_thread_ = std::make_unique<std::thread>(
+    [&]() {
+      executor_->spin();
+    });
+  spinner_thread_->detach();
+}
+
+Pause::~Pause()
+{
+  RCLCPP_DEBUG(node_->get_logger(), "Shutting down Pause BT node");
+  executor_->cancel();
+}
+
+BT::NodeStatus Pause::tick()
+{
+  unsigned int children_count = children_nodes_.size();
+  if (children_count < 1 || children_count > 4) {
```

Comment:
Docs on the `< 1 || > 4` is important. I'd add it both in the header doxygen, the BT walkthrough or tutorial, and the configuration guide page.